### ISSUE TITLE
Antivirus data stream problem fills log file

### ISSUE THREAD

**Issue by ms270169** [owncloud/files_antivirus#125]
In _OCA\Files_Antivirus\Scanner.initScanner()_ the nINSTREAM method is prepared to scan for a virus. In the clamav config file (_/etc/clamav/clamd.conf_) you can configure the maximum stream size with the line _StreamMaxLength 4095M_
With the value 4095M the maximum value of 4GiB is used, a bigger value is not supported by clamav.

Now in the function [OCA\Files_Antivirus\Scanner.scan(IScannable $item)](https://github.com/owncloud/files_antivirus/blob/master/lib/scanner.php#L61)  the scanning is done by cutting the file in small chunks (chunk size configureable, default size is 1024 bytes). Each chunk is streamed to the clamav-daemon.

If you have an video file with 5 GiB, there are 5368709120 chunks used. For the first 4294967296 chunks everything works fine.

But then the streaming size limit is reached and clamav-daemon closes the stream.
From this moment for each chunk the call to _fwrite()_ will throw an exception and cause a line in the log file.

For the example of a 5GiB video file, 1073741824 exceptions and log file lines will be created.
That needs a lot of time (several minutes), trashes full the log file and is a risc to running out of memory in the filesystem.

Correct behavior would be:
1) Not to use INSTREAM in case of file size over the stream limit
2) In case of exception to stop reading the file and only to log one error in the log file

**Comment by sidey79** [comment]
I've looket into this and found, that there can be a problem with async scans and background scans.

I've not understood all that wrapper things in owncloud, but i see, that the error is caused by fwrite.
The return code of fwrite is not checked, so fwrite is called for every chunk and it seems that it's not checked if fwrite has written all that data to the socket.

I've did a small quick and dirty extension to verify this:

```
public function onAsyncData($data){
                $rc=fwrite(
                                $this->getWriteHandle(),
                                $this->prepareChunk($data)
                );
                if ($rc === false || $rc == 0)
                {
                  $this->initAsyncScan();
                  fwrite(
                                $this->getWriteHandle(),
                                $this->prepareChunk($data)
                  );
                }

        }
```

With these additional lines of code, the rc of fwrite is checked. If data coul'd be written, a new socket will be opened. 

After that, i saw 3 entrys in the log, because fwrite may was able to send some of the 8192 bytes, but not all.
May this can be used as an idea, how to fix the problem or to quickly reduce the amount of logentrys.

**Comment by DeepDiver1975** [comment]
@VicDeo please have a look -THX

**Comment by felixboehm** [comment]
@VicDeo could you understand the issue details? How can we solve this?

**Comment by VicDeo** [comment]
@felixboehm yes. Can be solved by introducing the config setting identical to ClamAv `StreamMaxLength`  to prevent large file scanning (like most of sync-n-share solutions do)

**Comment by ms270169** [comment]
A config value can not be a "solution" for any situation, as long as it is possible that some files have more than 4GiB (see issue decription on top). A solution could be to use streaming as long as the file size does not exceed the ClamAv StreamMaxLength value, and to skip virus check otherwise or to use a different method (SCAN command, ...).

**Comment by VicDeo** [comment]
@ms270169 why not? 
1. Config option `Do not scan files larger than...`
2. Track amount of bytes passed to ClamAv
3. Consider closing a socket when this amount exceeds the option value to be normal and do not log anything in this case
4. Closing a socket when this amount do  not exceed the option value - log as before with a notice `check if the scanner option matches StreamMaxLength`
5. Adjust background scanner SQL query  to skip the files exceeding the option.

And maybe a button to  autosync this value with Clamav.

**Comment by sidey79** [comment]
What is with a option to open a new socket, if fwrite returns an error / steamsize Limit ist reached?

With this option it is possible to scan almost every size of files.

**Comment by VicDeo** [comment]
@sidey79 it makes the result unreliable. What if a virus signature will be split between these two chunks?

**Comment by sidey79** [comment]
@VicDeo i agree, but two things to note:

To not scan the entire file has a higher risk.
Even if there is a virus between the chunks it would be possible to scan some part from the previous chunk again. This is some overhead but could be specified with a config value like chunkRescanSize = xx KB


Otherwise, is there an option to scan those big files with a background process completly?

**Comment by ms270169** [comment]
@VicDeo your suggestion would solve the logging problem, but left the risk that infection of big files is not detected. Why not using alternatives if file size is too big. 
For example: The man page of clamd [https://linux.die.net/man/8/clamd](https://linux.die.net/man/8/clamd) offers the SCAN command.

I tested this SCAN command with `netcat -U /var/run/clamav/clamd.ctl` for a 5GiB file (filled with random bytes), and it results OK. But I am little bit confused about the execution time. OK was given immediatly after SCAN command was entered. Is it possible to scan a 5GiB file in a couple of ms?

**Comment by sidey79** [comment]
@ms270169 What die you Test with netcat?
I think you must specify the file on the file System which should be scanned.
There is also a config option for max scan size.
https://linux.die.net/man/5/clamd.conf

Not sure if this can be set to any size.

**Comment by ms270169** [comment]
@sidey79 You are right, I have overseen these parameters.
I have now set `MaxScanSize 10000M` and `MaxFileSize 10000M` in the file **/etc/clamav/clamd.conf**. After restarting the daemon, I connect with `netcat -U /var/run/clamav/clamd.ctl` and execute the command `SCAN /tmp/testfile`. The file **/tmp/testfile** is a 5GiB random byte file.
After entering the command I get immediately the message `/tmp/testfile: OK`.

But I am sure, the file was not really checked in this short time...

Also the man page description for these parameters is a little bit strange.
*Warning: disabling this limit or setting it too high may result in severe damage to the system.*
What should the be? What's the limit, what could happen?

So it seems, that using the SCAN command is not a workaround for the stream size limitation.

**Comment by VicDeo** [comment]
I think this is worth reading http://www.gossamer-threads.com/lists/clamav/users/67033#67033

**Comment by sidey79** [comment]
There are serval PHP Implementation for clamav.
They support the SCAN command which can be adjusted in size.

One option is also a php module:
http://php-clamav.sourceforge.net

**Comment by sidey79** [comment]
@ms270169
I found the following in the clamd.conf man Page:

Notes

All options expressing a size are limited to max 4GB. Values in excess will be resetted to the maximum. 

So ist seems that there is no option available to scan a file >4GB one part.

**Comment by VicDeo** [comment]
@sidey79 @ms270169 would you like to test https://github.com/owncloud/files_antivirus/issues/133?

**Comment by ms270169** [comment]
@VicDeo I have now tested [#133](https://github.com/owncloud/files_antivirus/pull/133).
In principle it works, but there are some things which should be improved in my point of view:

**1) config value AvMaxFileSize**
at the moment there is one config parameter *AvMaxFileSize* which is used (at least) two times. One usage is in [lib/backgroundscanner.php#L100](https://github.com/owncloud/files_antivirus/blob/e796ae3f0e53795b38da163a37637ef24d934175/lib/backgroundscanner.php#L100), another in [lib/scanner.php#L114](https://github.com/owncloud/files_antivirus/blob/e796ae3f0e53795b38da163a37637ef24d934175/lib/scanner.php#L114).  
I observe, that if the file is bigger than the config-limit, scanning is not executed, because the database query does not return this file (see backgroundscanner.php). So the new code lines in scanner.php are never reached.

I would suggest to introduce two config parameters. One config parameter *AvMaxFileSize* is used for the database query. If a file is above this limit, scanning is skipped. Normally this parameter is set to -1, which means there is no limit (as coded at the moment).

The second parameter *AvStreamMaxLength* (I take the name from clamav.config) is used in scanner.php to decide if scanner should be closed and reinitialized to prevent streaming problem.

**2) logging**
When I force streaming problem by wrong configuration (config limit higher than clamav StreamMaxLength parameter and file size between these two limits), I got a lot of log records in log file.
Your solution for reducing these log entries [lib/scanner.php#L122](https://github.com/owncloud/files_antivirus/blob/e796ae3f0e53795b38da163a37637ef24d934175/lib/scanner.php#L122) works, but call of *fwrite()* in line 120  also adds a log-record, and that happens many times (I have not found the location inside fwrite() where this log-record is written to log).

Each log-record looks like:
`{"reqId":"QpaPE1IicR\/hpupFjzYf","remoteAddr":"","app":"PHP","message":"fwrite(): send of 4 bytes failed with errno=32 Broken pipe at \/var\/www\/owncloud\/apps3\/files_antivirus\/lib\/scanner.php#120","level":3,"time":"2016-11-05T15:25:56+00:00","method":"--","url":"--","user":"admin"}`

But it's clear, if everything is configured proper, the problem should not occur. Maybe it would be a good idea to adjust the size limit if fwrite() fails and to report it in the log (admin would see that configuration is wrong), and to abort scanning if fwrite() fails a second time (because then there is something wrong with clamav daemon).

**3) call of fwrite() not checked**
In case of exceeded size limit the first call of fwrite [lib/scanner.php#L131](https://github.com/owncloud/files_antivirus/blob/e796ae3f0e53795b38da163a37637ef24d934175/lib/scanner.php#L131) does not check the return value. So it's intended that there is never a problem on first call of fwrite() after scanner reinitialization. It would be better to return to line 120 after reinitialization.

**4) block overlapping**
At the moment I cannot see any kind of overlapping blocks. So if a virus is in this overlapping area, it is probably not detected. As I see in [http://www.gossamer-threads.com/lists/clamav/users/67033#67033]( http://www.gossamer-threads.com/lists/clamav/users/67033#67033) (thanks for your comment, very interesting infos), block overlapping is not perfect, but better than nothing. Maybe I am wrong in my understanding of this problem...

That's all from my side at the moment. Thanks for taking this matter.

**Comment by sidey79** [comment]
With wath oc version can this version tested?
I've found the following in the branch:

`<owncloud min-version="9.2" max-version="9.2" />`

But i have no owncloud 9.2 installation.

**Comment by VicDeo** [comment]
@sidey79 I guess nobody has 9.2 atm (but Chuck Norris does)
What's your OC version?

**Comment by sidey79** [comment]
Wenn have 8.2 in Produktion und i have a 9.0 for some testing purpose.

VicDeo <notifications@github.com> schrieb am 08.11.2016:
>@sidey79 I guess nobody has 9.2 atm (but Chuck Norris does)
>What's your OC version?

-- Diese Nachricht wurde mit K-@ Mail gesendet.

**Comment by VicDeo** [comment]
@sidey79 yes, the PR above is not applicable to both.
When it is merged I prepare a backport.

**Comment by VicDeo** [comment]
@ms270169 
1) config value AvMaxFileSize
TBD

2) logging
"call of fwrite() in line 120"
native PHP logging suppressed

3) call of fwrite() not checked
"It would be better to return to line 120 after reinitialization."
it's a potential infinite loop. 

4) block overlapping
Technically possible. But I don't like the idea. 
Stream wrapper is a sort of callback that is called on file IO operation.
The implementation would be to store last successful chunk somewhere and retry with it first (and than with the current chunk) 
I can't explain why but I don't like it. :)

**Comment by VicDeo** [comment]
@ms270169 https://github.com/owncloud/files_antivirus/pull/133 updated according to your suggestions

**Comment by ms270169** [comment]
@VicDeo, great, works as desired. Test has shown that large files over config parameter *File size limit* are ignored, all other files are scanned. If stream size reaches config parameter *Stream Length* the scanner is reinitialized. Every file scanned without problem (also files above the clamav.config *StreamMaxLength* limit), no log entries visible (if log level is not debug).

I have seen, that you also implement overlapping blocks (obviously you changed your mind, and I am with you, because it makes code more complicate and it is not clear if it really increases virus detection probability ...) in  [scanner.php#L158](https://github.com/owncloud/files_antivirus/blob/c6c0f38ba6c3857274765ddc0c6448f3263e7cb5/lib/scanner.php#L158) (lastChunk). Maybe writing this lastChunk would also be needed after reinitialization in [L144](https://github.com/owncloud/files_antivirus/blob/c6c0f38ba6c3857274765ddc0c6448f3263e7cb5/lib/scanner.php#L144). I am not sure, but I guess that stream initialization leads to "restart" of the clamav detecting mechanism. In this case block overlapping would be a good idea (hopefully :-)

**Comment by VicDeo** [comment]
@ms270169 thanks for your in-depth analysis. Updated once again.

**Comment by ms270169** [comment]
@VicDeo, works fine with overlapping in case of reinitialization. I have seen, that you have removed the call of *ShutdownScanner()* in function *retry()* ([scanner.php#L165](https://github.com/owncloud/files_antivirus/blob/106c285c4c82d86f04d111f81b210ec6240f3643/lib/scanner.php#L165)). Is this correct? Because now you will not end with 'N' ([external.php#L44](https://github.com/owncloud/files_antivirus/blob/106c285c4c82d86f04d111f81b210ec6240f3643/lib/scanner/external.php#L44)), and therefor response of clamav-daemon is not handled (maybe I am wrong).

On recent versions I never checked, if in case of reinitialization response of clamav-daemon is taken and proper handled. Is a file infection detected if the first part of file is infected by a virus? By viewing code I am not sure if response is proper handled. I will try to get a virus test signature and test again today or tomorrow with such a file to see if it works when clamav detects something.

**Comment by ms270169** [comment]
@VicDeo. I have finished my tests with the following setup:

* testfile: 40000 bytes length, 68 byte EICAR virus signature on position 21000-21067 (would only be detected if block overlapping works)
* config-value *Chunk Size*: 8192 (same value as hard coded in [helper.php#L291](https://github.com/owncloud/core/blob/7fc4642af25bc90781ee28e294511b75f6e0fbe6/lib/private/legacy/helper.php#L291]))
* config-value *Stream Length*: 20000
* config-value *File size limit*: -1

I add the 40000 byte testfile from the file menu:
Desired result: virus detected and reported to the user

Result from Test:
* on current version (106c285): virus not detected, file uploaded without message
* on recent version (c6c0f38b): "Unsupported Media Type", file not uploaded

**Reasons for recent version (c6c0f38b) behavior**
After reaching the stream limit, the function *shutdownScanner*([scanner.php#L143](https://github.com/owncloud/files_antivirus/blob/c6c0f38ba6c3857274765ddc0c6448f3263e7cb5/lib/scanner.php#L143)) is called. There the "N" is sent to clamav-daemon and the scan-result is proper handled. The EICAR virus signature is detected and a *InvalidContentException* is thrown ([avirwrapper.php#L99](https://github.com/owncloud/files_antivirus/blob/c6c0f38ba6c3857274765ddc0c6448f3263e7cb5/lib/avirwrapper.php#L99)). Up to now everything is right.
Now the exception is catched in [File.php#L153](https://github.com/owncloud/core/blob/7fc4642af25bc90781ee28e294511b75f6e0fbe6/apps/dav/lib/Connector/Sabre/File.php#L153) and converted into a *SabreException* in the function *convertToSabreException*. And here the exception is converted into a *UnsupportedMediaType* ([File.php#L535](https://github.com/owncloud/core/blob/7fc4642af25bc90781ee28e294511b75f6e0fbe6/apps/dav/lib/Connector/Sabre/File.php#L535)), what is definitely a bug. The exception message is correct up to now (*Virus Eicar-Test-Signature is detected in the file. Upload cannot be completed.*), but the message is not shown to the user. So the user will ask himself what happens, whats wrong with the media type? No hint to a virus. No record in log.

**Reasons for current version (c6c0f38b) behavior**
After reaching the stream limit, the stream is reinitialized. A shutdown before is not done. So the status of the clamav-daemon is neither requested (nor handled).

**Additional Topic**
Up to now I test the scanner via the scanner job. Now I have tested the scanner via the "File Add" button. I wondered first why the chunk size is 8192 byte instead of my recent antivirus config value 1024. I found the answer in [helper.php#L291](https://github.com/owncloud/core/blob/7fc4642af25bc90781ee28e294511b75f6e0fbe6/lib/private/legacy/helper.php#L291). There you can see, the value is hard coded into source. If it is not possible to change this behavior, I would recommend to remove the config value for chunk size and to take over the same value (8192).

**Comment by VicDeo** [comment]
@ms270169 
>why the chunk size is 8192 byte instead of my recent antivirus config value 1024. I found the answer in helper.php#L291

This code mimics  [ an old PHP bug](http://php.net/manual/en/function.stream-wrapper-register.php#74765), but the chunk size may vary

**Comment by VicDeo** [comment]
@ms270169 I updated PR once again. Could you please it another try and review it dirrectly in PR?

**Comment by ManfredSteiner** [comment]
Conversation now changed to [https://github.com/owncloud/files_antivirus/pull/133](https://github.com/owncloud/files_antivirus/pull/133). (Hopefully the right one PR)

**Comment by ms270169** [comment]
Conversation for test of fixes now changed to [#133](https://github.com/owncloud/files_antivirus/pull/133#issuecomment-260854367). (Hopefully the right one PR)

**Comment by PVince81** [comment]
PR was merged, closing
